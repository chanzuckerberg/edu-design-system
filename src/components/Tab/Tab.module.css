@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    #TAB
\*------------------------------------*/
