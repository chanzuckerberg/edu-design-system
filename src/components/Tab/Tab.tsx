import React, { ReactNode } from 'react';

export interface Props {
  /**
   * aria-labelledby attribute that associates a tab panel with its accompanying tab title text
   */
  'aria-labelledby'?: string;
  /**
   * Child node(s) that can be nested inside component
   */
  children?: ReactNode;
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
  /**
   * HTML id for the component
   */
  id?: string;
  /**
   * Number passed down from Tabs to show the active index state of Tabs
   */
  isActiveIndex?: number;
  /**
   * Number passed down from Tabs to show the active index state of Tabs
   */
  title?: string;
}

/**
 * ```ts
 * import {Tab} from "@chanzuckerberg/eds";
 * ```
 *
 * Individual tab within the Tabs component.
 */
export const Tab = ({
  children,
  className,
  id,
  'aria-labelledby': ariaLabelledBy,
  title,
  ...other
}: Props) => {
  return (
    <div
      aria-hidden={false}
      aria-labelledby={ariaLabelledBy}
      className={className}
      id={id}
      role="tabpanel"
      {...other}
    >
      {children}
    </div>
  );
};
