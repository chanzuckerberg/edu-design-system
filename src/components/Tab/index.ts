export { Tab as default } from './Tab';
export type { Props as TabProps } from './Tab';
