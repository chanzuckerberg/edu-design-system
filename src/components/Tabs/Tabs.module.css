@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    #TABS
\*------------------------------------*/

/**
 * List of of links where each link toggles open associated information
 */

/**
 * Tabs header
 *
 * Contains the tab list.
 */
.tabs__header {
  /* Tab overflow behavior*/
  overflow-x: auto;
  position: relative;
  padding-bottom: var(--eds-size-2);
}

/**
 * Tabs list
 *
 * Actual unordered list of tabs.
 */
.tabs__list {
  @mixin reset;
  list-style: none;
  /* Set to display flex to place tabs side by side. */
  display: flex;
  border-bottom: var(--eds-theme-border-width) solid
    var(--eds-theme-color-border-neutral-subtle);
  @mixin eds-theme-typography-tab;
}

/**
 * Tabs item
 */
.tabs__item {
  /**
   * Flex shrink 0 keeps tabs from expanding to fill up the space of the container.
   */
  flex-shrink: 0;
}

/**
 * Tabs link
 */
.tabs__link {
  display: block;
  padding: var(--eds-size-1-and-half);
  text-decoration: none;
  color: var(--eds-theme-color-text-neutral-default);
  transition: color var(--eds-anim-fade-quick) var(--eds-anim-ease),
    box-shadow var(--eds-anim-fade-quick) var(--eds-anim-ease),
    background-color var(--eds-anim-fade-quick) var(--eds-anim-ease);

  &:focus-visible {
    @mixin focus;
  }

  @supports not selector(:focus-visible) {
    &:focus {
      @mixin focus;
    }
  }

  @media screen and (prefers-reduced-motion) {
    transition: none;
  }

  /**
   * Tabs link within active tabs item
   */
  .tabs__item.eds-is-active & {
    color: var(--eds-theme-color-text-link-brand);
    box-shadow: inset 0 calc(var(--eds-border-width-md) * -1) 0 0
      var(--eds-theme-color-border-brand-primary-strong);
  }
}
