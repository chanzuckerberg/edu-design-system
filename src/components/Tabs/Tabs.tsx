import clsx from 'clsx';
import debounce from 'lodash.debounce';
import React, {
  type ReactNode,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
  type KeyboardEvent,
} from 'react';
import { allByType } from 'react-children-by-type';
import { useUID, useUIDSeed } from 'react-uid';
import styles from './Tabs.module.css';
import {
  L_ARROW_KEYCODE,
  U_ARROW_KEYCODE,
  R_ARROW_KEYCODE,
  D_ARROW_KEYCODE,
} from '../../util/keycodes';
import Tab from '../Tab';

export interface Props {
  /**
   * Reference to another element that describes the purpose of the set of tabs.
   */
  'aria-labelledby'?: string;
  /**
   * Calls back with the active index
   */
  onChange?: (index: number) => void;
  /**
   * Child node(s) that can be nested inside component
   */
  children?: ReactNode;
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
  /**
   * HTML id for the component
   */
  id?: string;
  /**
   * Passed down to initially set the activeIndex state
   */
  activeIndex?: number;
}

/**
 * ```ts
 * import {Tabs} from "@chanzuckerberg/eds";
 * ```
 *
 * List of of links where each link toggles open associated information in a tab panel.
 */
export const Tabs = ({
  activeIndex = 0,
  'aria-labelledby': ariaLabelledBy,
  children,
  className,
  onChange,
  ...other
}: Props) => {
  const getUID = useUIDSeed();
  const activeTabPanelId = useUID();
  const headerRef = useRef<HTMLDivElement>(null);
  const [activeIndexState, setActiveIndexState] = useState(activeIndex);
  const [scrollableLeft, setScrollableLeft] = useState<boolean>(false);
  const [scrollableRight, setScrollableRight] = useState<boolean>(false);

  /** Children that are actually tabs. Any others are ignored. */
  const tabs = useMemo(() => {
    return allByType(children, Tab);
  }, [children]);

  const tabRefs = useMemo(
    () => tabs.map(() => React.createRef<HTMLAnchorElement>()),
    [tabs],
  );

  const tabIds = useMemo(
    () => tabs.map((tab) => tab.props.id || getUID(tab)),
    [tabs, getUID],
  );

  // Set the active tab if the `activeIndex` prop changes.
  const prevActiveIndex = usePrevious(activeIndex);
  useEffect(() => {
    if (prevActiveIndex != null && prevActiveIndex !== activeIndex) {
      setActiveIndexState(activeIndex);
      tabRefs[activeIndex].current?.focus();
    }
  }, [prevActiveIndex, activeIndex, tabRefs]);

  /**
   * Handles if scroll fade indicators should be displayed.
   */
  const handleTabsScroll = useCallback((headerEl: HTMLDivElement) => {
    const scrollLeft = headerEl.scrollLeft;
    const width = headerEl.clientWidth;
    const scrollWidth = headerEl.scrollWidth;

    if (scrollLeft > 0) {
      setScrollableLeft(true);
    } else {
      setScrollableLeft(false);
    }

    if (scrollWidth > width && scrollLeft + width < scrollWidth) {
      setScrollableRight(true);
    } else {
      setScrollableRight(false);
    }
  }, []);

  /**
   * Listens for window resize to display scroll fade indicators.
   */
  useEffect(() => {
    if (headerRef && headerRef.current) {
      const resizeHandleTabs = debounce(
        () => {
          if (headerRef.current) {
            handleTabsScroll(headerRef.current);
          }
        },
        100,
        { leading: true },
      );

      /**
       * The event listener actually calls the callback once when initiated, but the event listener
       * is not triggered with prop changes so this line is required.
       * This means the callback may be called twice on initial paint, which is fine, and
       * is better than it not being called at all.
       */
      resizeHandleTabs();
      window.addEventListener('resize', resizeHandleTabs);
      return () => {
        window.removeEventListener('resize', resizeHandleTabs);
      };
    }
  }, [handleTabsScroll]);

  function handleClick(index: number) {
    setActiveIndexState(index);

    if (onChange) {
      onChange(index);
    }
  }

  function handleKeyDown(e: KeyboardEvent<HTMLAnchorElement>) {
    let activeTab = null;

    tabRefs.map((item) => {
      if (item.current === document.activeElement) {
        activeTab = item;
      }
      return item;
    });

    if (!activeTab) return;

    // Set active, next, and previous tab.
    const index = tabRefs.indexOf(activeTab);
    const next = index === tabRefs.length - 1 ? 0 : index + 1;
    const prev = index === 0 ? tabRefs.length - 1 : index - 1;

    if ([R_ARROW_KEYCODE, D_ARROW_KEYCODE].includes(e.key)) {
      // Right or down arrow was pressed. Focus the next tab.
      tabRefs[next].current?.focus();
    } else if ([L_ARROW_KEYCODE, U_ARROW_KEYCODE].includes(e.key)) {
      // Left or up was pressed. Focus the previous tab.
      tabRefs[prev].current?.focus();
    }
  }

  const componentClassName = clsx(styles['tabs'], className);

  const headerClassName = clsx(
    styles['tabs__header'],
    scrollableLeft && styles['tabs--scrollable-left'],
    scrollableRight && styles['tabs--scrollable-right'],
  );

  const activeTabPanel = React.cloneElement(tabs[activeIndexState], {
    id: activeTabPanelId,
    'aria-labelledby': tabIds[activeIndexState],
  });

  return (
    <div className={componentClassName} {...other}>
      <div
        className={headerClassName}
        onScroll={(e) => handleTabsScroll(e.target as HTMLDivElement)}
        ref={headerRef}
      >
        <ul
          aria-labelledby={ariaLabelledBy}
          className={styles['tabs__list']}
          role="tablist"
        >
          {tabs.map((tab, i) => {
            const isActive = activeIndexState === i;
            return (
              <li
                className={clsx(
                  styles['tabs__item'],
                  isActive && styles['eds-is-active'],
                )}
                key={tabIds[i]}
                role="presentation"
              >
                <a
                  aria-controls={activeTabPanelId}
                  aria-selected={isActive}
                  className={styles['tabs__link']}
                  href={`#${activeTabPanelId}`}
                  id={tabIds[i]}
                  onClick={(e) => {
                    e.preventDefault();
                    handleClick(i);
                  }}
                  onKeyDown={handleKeyDown}
                  ref={tabRefs[i]}
                  role="tab"
                  tabIndex={isActive ? 0 : -1}
                >
                  {tab.props.title}
                </a>
              </li>
            );
          })}
        </ul>
      </div>
      <div>{activeTabPanel}</div>
    </div>
  );
};

/**
 * Get the previous value of a prop. Useful for comparing the previous to the current value.
 */
function usePrevious<T>(prop: T) {
  const ref = useRef<T>(prop);
  useEffect(() => {
    ref.current = prop;
  });
  return ref.current;
}
