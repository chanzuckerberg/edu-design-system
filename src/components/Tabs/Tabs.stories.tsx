import type { StoryObj, Meta } from '@storybook/react';
import { within } from '@storybook/testing-library';
import React from 'react';

import { Tabs } from './Tabs';
import { chromaticViewports } from '../../util/viewports';
import Heading from '../Heading';
import Icon, { type IconName } from '../Icon';
import Text from '../Text';

export default {
  title: 'Components/Tabs',
  component: Tabs,
  parameters: {
    layout: 'centered',
    badges: ['intro-1.0'],
  },
  args: {
    children: (
      <>
        <Tabs.Tab title="Tab Title 1">
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 1
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 2">
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 2
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 3">
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 3
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 4">
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 4
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 5">
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 5
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 6">
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 6
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 7">
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 7
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 8">
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 8
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 9">
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 9
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>
      </>
    ),
  },
  argTypes: {
    children: {
      control: {
        type: null,
      },
    },
  },
} as Meta<Args>;

type Args = React.ComponentProps<typeof Tabs>;

export const Default: StoryObj<Args> = {
  parameters: {
    chromatic: {
      viewports: [
        chromaticViewports.googlePixel2,
        chromaticViewports.chromebook,
      ],
    },
  },
};

// For visual regression testing of both masks
export const ScrollMiddle: StoryObj<Args> = {
  parameters: {
    viewport: {
      defaultViewport: 'googlePixel2',
    },
    // Skip these b/c test environment cannot execute "scroll" on the parent div
    snapshot: { skip: true },
    chromatic: { viewports: [chromaticViewports.googlePixel2] },
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const tablist = await canvas.findByRole('tablist');
    // eslint-disable-next-line testing-library/no-node-access
    tablist?.parentElement?.scroll(50, 0);
  },
  decorators: [
    (Story) => (
      <div>
        For Chromatic visual regression testing of the masks on both sides of
        the Tabs. Currently does not work properly on local.
        {Story()}
      </div>
    ),
  ],
};

export const Centered: StoryObj<Args> = {
  args: {
    align: 'center',
  },
  parameters: {
    chromatic: {
      viewports: [
        chromaticViewports.googlePixel2,
        chromaticViewports.chromebook,
      ],
    },
  },
};

const IconTab = ({
  isActive,
  title,
  icon,
}: {
  isActive: boolean;
  title: string;
  icon: IconName;
}) => (
  <div className="px-4">
    <div className="p-3 text-center">
      <Icon name={icon} purpose="decorative" size="1rem" />
    </div>
    <Text preset="caption-lg">{`${isActive ? '●' : '◦'} ${title}`}</Text>
  </div>
);

export const CustomTabs: StoryObj<Args> = {
  decorators: [(Story) => <div className="p-8">{Story()}</div>],
  parameters: {
    docs: {
      source: {
        code: `
<Tabs>
  <Tabs.Tab title="Tab Title 1">
    <Tabs.Tab.Button>
      {({ active, title }) => (
        <IconTab
          icon={active ? 'person' : 'people'}
          isActive={active}
          title={title}
        />
      )}
    </Tabs.Tab.Button>
    <div className="max-w-xl">
      <Heading className="mb-6" size="h3">
        Tab 1
      </Heading>
      <Text>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
        eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
        enim ad minim veniam, quis nostrud exercitation ullamco laboris
        nisi ut aliquip ex{' '}
      </Text>
    </div>
  </Tabs.Tab>

  <Tabs.Tab title="Tab Title 2">
    <Tabs.Tab.Button>
      {({ active, title }) => (
        <IconTab
          icon={active ? 'person' : 'people'}
          isActive={active}
          title={title}
        />
      )}
    </Tabs.Tab.Button>
    <div className="max-w-xl">
      <Heading className="mb-6" size="h3">
        Tab 2
      </Heading>
      <Text>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
        eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
        enim ad minim veniam, quis nostrud exercitation ullamco laboris
        nisi ut aliquip ex{' '}
      </Text>
    </div>
  </Tabs.Tab>

  <Tabs.Tab title="Tab Title 3">
    <Tabs.Tab.Button>
      {({ active, title }) => (
        <IconTab
          icon={active ? 'person' : 'people'}
          isActive={active}
          title={title}
        />
      )}
    </Tabs.Tab.Button>
    <div className="max-w-xl">
      <Heading className="mb-6" size="h3">
        Tab 3
      </Heading>
      <Text>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
        eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
        enim ad minim veniam, quis nostrud exercitation ullamco laboris
        nisi ut aliquip ex{' '}
      </Text>
    </div>
  </Tabs.Tab>

  <Tabs.Tab title="Tab Title 4">
    <Tabs.Tab.Button>
      {({ active, title }) => (
        <IconTab
          icon={active ? 'person' : 'people'}
          isActive={active}
          title={title}
        />
      )}
    </Tabs.Tab.Button>
    <div className="max-w-xl">
      <Heading className="mb-6" size="h3">
        Tab 4
      </Heading>
      <Text>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
        eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
        enim ad minim veniam, quis nostrud exercitation ullamco laboris
        nisi ut aliquip ex{' '}
      </Text>
    </div>
  </Tabs.Tab>

  <Tabs.Tab title="Tab Title 5">
    <Tabs.Tab.Button>
      {({ active, title }) => (
        <IconTab
          icon={active ? 'person' : 'people'}
          isActive={active}
          title={title}
        />
      )}
    </Tabs.Tab.Button>
    <div className="max-w-xl">
      <Heading className="mb-6" size="h3">
        Tab 5
      </Heading>
      <Text>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
        eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
        enim ad minim veniam, quis nostrud exercitation ullamco laboris
        nisi ut aliquip ex{' '}
      </Text>
    </div>
  </Tabs.Tab>

  <Tabs.Tab title="Tab Title 6">
    <Tabs.Tab.Button>
      {({ active, title }) => (
        <IconTab
          icon={active ? 'person' : 'people'}
          isActive={active}
          title={title}
        />
      )}
    </Tabs.Tab.Button>
    <div className="max-w-xl">
      <Heading className="mb-6" size="h3">
        Tab 6
      </Heading>
      <Text>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
        eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
        enim ad minim veniam, quis nostrud exercitation ullamco laboris
        nisi ut aliquip ex{' '}
      </Text>
    </div>
  </Tabs.Tab>
</Tabs>
        `,
      },
    },
  },
  args: {
    children: (
      <>
        <Tabs.Tab title="Tab Title 1">
          <Tabs.Tab.Button>
            {({ active, title }) => (
              <IconTab
                icon={active ? 'person' : 'people'}
                isActive={active}
                title={title}
              />
            )}
          </Tabs.Tab.Button>
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 1
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 2">
          <Tabs.Tab.Button>
            {({ active, title }) => (
              <IconTab
                icon={active ? 'person' : 'people'}
                isActive={active}
                title={title}
              />
            )}
          </Tabs.Tab.Button>
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 2
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 3">
          <Tabs.Tab.Button>
            {({ active, title }) => (
              <IconTab
                icon={active ? 'person' : 'people'}
                isActive={active}
                title={title}
              />
            )}
          </Tabs.Tab.Button>
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 3
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 4">
          <Tabs.Tab.Button>
            {({ active, title }) => (
              <IconTab
                icon={active ? 'person' : 'people'}
                isActive={active}
                title={title}
              />
            )}
          </Tabs.Tab.Button>
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 4
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 5">
          <Tabs.Tab.Button>
            {({ active, title }) => (
              <IconTab
                icon={active ? 'person' : 'people'}
                isActive={active}
                title={title}
              />
            )}
          </Tabs.Tab.Button>
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 5
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>

        <Tabs.Tab title="Tab Title 6">
          <Tabs.Tab.Button>
            {({ active, title }) => (
              <IconTab
                icon={active ? 'person' : 'people'}
                isActive={active}
                title={title}
              />
            )}
          </Tabs.Tab.Button>
          <div className="max-w-xl">
            <Heading as="h3" className="mb-6">
              Tab 6
            </Heading>
            <Text>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut
              enim ad minim veniam, quis nostrud exercitation ullamco laboris
              nisi ut aliquip ex{' '}
            </Text>
          </div>
        </Tabs.Tab>
      </>
    ),
  },
};
