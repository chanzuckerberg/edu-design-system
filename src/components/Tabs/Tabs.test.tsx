import { generateSnapshots } from '@chanzuckerberg/story-utils';
import { composeStories } from '@storybook/testing-react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import * as stories from './Tabs.stories';
import Tab from '../Tab';

const { Default } = composeStories(stories);

describe('<Tabs />', () => {
  generateSnapshots(stories);

  it('should focus and select with keyboard controls', async () => {
    const user = userEvent.setup();
    render(<Default />);
    const firstTab = screen.getByRole('tab', { name: 'Tab Title 1' });
    const secondTab = screen.getByRole('tab', { name: 'Tab Title 2' });
    firstTab.focus();

    await user.keyboard('{arrowright}');
    expect(secondTab).toHaveFocus();

    await user.keyboard('{arrowleft}');
    expect(firstTab).toHaveFocus();

    await user.keyboard('{arrowdown}');
    expect(secondTab).toHaveFocus();

    await user.keyboard('{arrowup}');
    expect(firstTab).toHaveFocus();

    await user.keyboard('{arrowdown}');
    await user.keyboard('{arrowright}');
    await user.keyboard('{enter}');
    expect(screen.getByRole('tab', { name: 'Tab Title 3' })).toHaveAttribute(
      'aria-selected',
      'true',
    );
  });

  it('changes the active tab when activeIndex changes', () => {
    const { rerender } = render(<Default activeIndex={0} />);
    expect(screen.getByRole('heading', { name: 'Tab 1' })).toBeInTheDocument();

    rerender(<Default activeIndex={1} />);
    expect(screen.getByRole('heading', { name: 'Tab 2' })).toBeInTheDocument();
  });

  it('does not include invalid characters in tab ids', () => {
    render(
      <Default id="foo">
        <Tab data-testid="tab-1" title="Tab Title 1">
          Tab numero uno
        </Tab>
      </Default>,
    );

    expect(screen.getByTestId('tab-1')).toHaveAttribute(
      'aria-labelledby',
      'foo-Tab-Title-1',
    );
  });
});
