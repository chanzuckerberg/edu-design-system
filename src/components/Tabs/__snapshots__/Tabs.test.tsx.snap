// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> Default story renders snapshot 1`] = `
<div
  class="tabs"
>
  <div
    class="tabs__header"
  >
    <ul
      class="tabs__list"
      role="tablist"
    >
      <li
        class="tabs__item eds-is-active"
        role="presentation"
      >
        <a
          aria-controls=":r0:"
          aria-selected="true"
          class="tabs__link"
          href="#:r0:"
          id=":r1:-Tab Title 1"
          role="tab"
          tabindex="0"
        >
          Tab Title 1
        </a>
      </li>
      <li
        class="tabs__item"
        role="presentation"
      >
        <a
          aria-controls=":r0:"
          aria-selected="false"
          class="tabs__link"
          href="#:r0:"
          id=":r1:-Tab Title 2"
          role="tab"
          tabindex="-1"
        >
          Tab Title 2
        </a>
      </li>
      <li
        class="tabs__item"
        role="presentation"
      >
        <a
          aria-controls=":r0:"
          aria-selected="false"
          class="tabs__link"
          href="#:r0:"
          id=":r1:-Tab Title 3"
          role="tab"
          tabindex="-1"
        >
          Tab Title 3
        </a>
      </li>
      <li
        class="tabs__item"
        role="presentation"
      >
        <a
          aria-controls=":r0:"
          aria-selected="false"
          class="tabs__link"
          href="#:r0:"
          id=":r1:-Tab Title 4"
          role="tab"
          tabindex="-1"
        >
          Tab Title 4
        </a>
      </li>
      <li
        class="tabs__item"
        role="presentation"
      >
        <a
          aria-controls=":r0:"
          aria-selected="false"
          class="tabs__link"
          href="#:r0:"
          id=":r1:-Tab Title 5"
          role="tab"
          tabindex="-1"
        >
          Tab Title 5
        </a>
      </li>
      <li
        class="tabs__item"
        role="presentation"
      >
        <a
          aria-controls=":r0:"
          aria-selected="false"
          class="tabs__link"
          href="#:r0:"
          id=":r1:-Tab Title 6"
          role="tab"
          tabindex="-1"
        >
          Tab Title 6
        </a>
      </li>
      <li
        class="tabs__item"
        role="presentation"
      >
        <a
          aria-controls=":r0:"
          aria-selected="false"
          class="tabs__link"
          href="#:r0:"
          id=":r1:-Tab Title 7"
          role="tab"
          tabindex="-1"
        >
          Tab Title 7
        </a>
      </li>
      <li
        class="tabs__item"
        role="presentation"
      >
        <a
          aria-controls=":r0:"
          aria-selected="false"
          class="tabs__link"
          href="#:r0:"
          id=":r1:-Tab Title 8"
          role="tab"
          tabindex="-1"
        >
          Tab Title 8
        </a>
      </li>
      <li
        class="tabs__item"
        role="presentation"
      >
        <a
          aria-controls=":r0:"
          aria-selected="false"
          class="tabs__link"
          href="#:r0:"
          id=":r1:-Tab Title 9"
          role="tab"
          tabindex="-1"
        >
          Tab Title 9
        </a>
      </li>
    </ul>
  </div>
  <div>
    <div
      aria-hidden="false"
      aria-labelledby=":r1:-Tab Title 1"
      id=":r0:"
      role="tabpanel"
    >
      <div
        class="max-w-xl"
      >
        <h3
          class="heading heading--size-h3 mb-6"
        >
          Tab 1
        </h3>
        <p
          class="text text--body"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex
           
        </p>
      </div>
    </div>
  </div>
</div>
`;
